src/Proto.Actor/Extensions/ActorSystemExtensions.cs

```diff
@@ -0,0 +1,28 @@
+using System.Collections.Generic;
+
+namespace Proto.Extensions
+{
+    public class ActorSystemExtensions
+    {
+        private readonly IActorSystemExtension[] _extensions = new IActorSystemExtension[100];
+        private readonly ActorSystem _actorSystem;
+
+        public ActorSystemExtensions(ActorSystem actorSystem)
+        {
+            _actorSystem = actorSystem;
+        }
+
+        public T Get<T>() where T: IActorSystemExtension
+        {
+            var id = IActorSystemExtension<T>.Id;
+            return (T)_extensions[id];
+        }
+
+        public void Register<T>(IActorSystemExtension<T> extension) where T:IActorSystemExtension
+        {
+            var id = IActorSystemExtension<T>.Id;
+            //TODO, ensure capacity
+            _extensions[id] = extension;
```

Comment:
I don't think you need to optimize here when adding extensions. Just replace the _extensions array with a bigger one with all the old ones copied over. Then there is no set limit to how many plugins one can add, and you can use a smaller default size..

Reply:
Oh, and ofc lock when switching out the array :)